src/coreclr/nativeaot/Runtime/gcrhenv.cpp

```diff
@@ -292,13 +293,32 @@ Object* GcAllocInternal(MethodTable *pEEType, uint32_t uFlags, uintptr_t numElem
 //  pEEType         -  type of the object
 //  uFlags          -  GC type flags (see gc.h GC_ALLOC_*)
 //  numElements     -  number of array elements
-//  pTransitionFrame-  transition frame to make stack crawable
+//  pTransitionFrame-  transition frame to make stack crawlable
 // Returns a pointer to the object allocated or NULL on failure.
 
 COOP_PINVOKE_HELPER(void*, RhpGcAlloc, (MethodTable* pEEType, uint32_t uFlags, uintptr_t numElements, PInvokeTransitionFrame* pTransitionFrame))
 {
     Thread* pThread = ThreadStore::GetCurrentThread();
 
+    // The allocation fast path is an asm helper that runs in coop mode and handles most allocation cases.
+    // The helper can also be tail-called. That is desirable for the fast path.
+    //
+    // Here we are on the slow(er) path when we need to call into GC. The fast path pushes a frame and calls here.
+    // In extremely rare cases the caller of the asm helper is hijacked and the helper is tail-called.
+    // As a result the asm helper may capture a hijacked return address into the transition frame.
+    // We do not want to put the burden of preventing such scenario on the fast path. Instead we will
+    // check for "hijacked frame" here and un-hijack m_RIP.
+    // We do not need to re-hijack when we are done, since m_RIP is discarded in POP_COOP_PINVOKE_FRAME
+    //
+    // NOTE: this is not a problem for ARM since tail-calling methods are not hijackable for other reasons.
```

Comment:
Is there any harm in enabling this for all platforms? This is a very cheap check.

Reply:
It is hard to reason how we would deal with this issue on ARM64. 

m_RIP is not discarded in POP_COOP_PINVOKE_FRAME` on architectures that have LR-style calls. POP_COOP_PINVOKE_FRAME would use m_RIP to restore LR.  
The contract with hijacking is that we can never pop out of a hijacked frame. Otherwise unhijack would be modifying stack that could be reused in subsequent calls for something else. So, if LR somehow contained a hijacked return, we'd need to restore it.

I figured it is easier to not do the check at all and explain why it is not needed, than do the check and explain why the fixup might not work correctly on ARM64 but also would never be applied.

Reply:
Another question here is - why do we not unhijack the thread as well? 

We could do that. It is allowed for a thread in coop mode to unhijack itself at any time. But it is more work and while this is a slower path, we are not necessarily heading for a GC. Most likely GC will just refill/reset the allocation buffer minus the object that still needs to be allocated.

It can happen relatively fast and hijack would still be valid. So I figured to do the minimum fixup that is needed.

Reply:
It was not clear to me that this logic would be broken on Arm64. 

The norm on newer architectures is to have the return address in a register. Amd64 and x86 are the exception. 

I think we should:

- Put this logic under `#if defined(TARGET_X86) || defined(TARGET_AMD64)`
- Add `ASSERT(!Thread::IsHijackTarget(pTransitionFrame->m_RIP));` into the `#else` part of the ifdef, with a comment explaining the problem.

Reply:
> The norm on newer architectures is to have the return address in a register. Amd64 and x86 are the exception.

Yes, I think new architectures will use something like LR. Another reason I did not want this to be a default. 

> Put this logic under #if defined(TARGET_X86) || defined(TARGET_AMD64)

Right. It would be better.

> Add ASSERT(!Thread::IsHijackTarget(pTransitionFrame->m_RIP)); into the #else part of the ifdef, with a comment explaining the problem.

Can add that.

I've added such assert to the `SetDeferredTransitionFrame` - to catch all the cases - where this condition cannot happen or where it needs a fixup.

In theory we could introduce more cases that would call `SetDeferredTransitionFrame` and have a similar concern. For now I think RhpGcAlloc is the only such case.

Reply:
There is one another user of PUSH_COOP_PINVOKE_FRAME - RhpGcPoll which is used for polling in loops. I do not think it can be tail-called, since it is called from loops.

I think, I will add an assert in RhpGcPoll2, just in case.